### PR TITLE
Add CI/testing to examples

### PR DESCRIPTION
**Pull request by phlax**
Commit Message: Add CI/testing to examples
Additional Description:

This PR incorporates the tests that i created for the envoy examples (originally here: https://github.com/phlax/envoy-examples)

It also adds a test to ensure that example configs have been added to the examples `BUILD` config (ref: https://github.com/envoyproxy/envoy/pull/12256#discussion_r464512574)

the tests are documented in #12874 

## original description - no longer accurate...
the tests are run inside a `docker-in-docker` container to ensure a clean environment, and to allow docker system pruning without affecting the hosts container setup.

The container user's `umask` is set to `027` to catch any potential permissions problems related to running envoy as non-root in containers

The tests can be run locally with `./ci/verify_examples.sh` from the root of the envoy repo. This will mount the current directory as the envoy source

Alternatively you can do eg the following to test a different branch/source:

```
export ENVOY_REPO="-b some-other-branch https://github.com/phlax/envoy"
./ci/verify_examples.sh
```
In this case the build config test is skipped

Risk Level: low
Testing: yes
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by phlax** [comment]
@dio @lizan ive created this PR to add the example tests i created, and hoping it would trigger in ci etc

 i added a separate job in azure pipelines but that doesnt seem to have worked.

im wondering if that is the best way to trigger this in ci - and if so - whether something needs to be added in azure to allow this job to run

File: .azure-pipelines/pipelines.yml
Comment (lizan) [inline .azure-pipelines/pipelines.yml:45]: I'm OK do this as part of existing docker job, so you don't have to wait for the image built.

File: ci/run_example_tests.sh
Comment (lizan) [inline ci/run_example_tests.sh:462]: Is there a way to run examples without listing examples here?

File: ci/run_example_tests.sh
Comment (lizan) [inline ci/run_example_tests.sh:49]: You shouldn't do this, use CI checked-out one.

File: ci/run_example_tests.sh
Comment (phlax) [reply ci/run_example_tests.sh:462]: yep, i can update

File: ci/run_example_tests.sh
Comment (phlax) [reply ci/run_example_tests.sh:49]: it only does this if the already checked out repo isnt mounted - this allows it to be used against other branches

i can remove, but its kinda useful

File: .azure-pipelines/pipelines.yml
Comment (phlax) [reply .azure-pipelines/pipelines.yml:45]: im not totally following here, couple of questions

- whereabouts in ci is best to run this job ?
- are you happy with running these tests with `docker-in-docker` ?

Comment (phlax) [reply .azure-pipelines/pipelines.yml:45]: i guess theres another question about the envoy container that is being tested in the examples - atm it would use the published one - ie not the one which would be built from the tested branch

Comment (lizan) [reply .azure-pipelines/pipelines.yml:45]: I'm saying run this as a part of https://github.com/envoyproxy/envoy/blob/master/ci/docker_ci.sh

We used to have a TODO for that and lost during refactoring: https://github.com/envoyproxy/envoy/blob/v1.13.0/ci/docker_push.sh#L24-L27

File: ci/run_example_tests.sh
Comment (lizan) [reply ci/run_example_tests.sh:49]: If you want to check against other branch, just checkout it? Are you saying you want to check branches other than where envoy is built from? Then the argument should be the docker image tag.

Comment (phlax) [reply ci/run_example_tests.sh:49]: i guess mostly because ive been running this in an environment that does not already have a checkout of the envoy repo (ie on travis) it has been much easier to do this:

```
export ENVOY_REPO=...

docker run --rm phlax/envoy-examples 

```

rather than having to checkout the repo and mount it into the container

im happy to remove

Comment (lizan) [reply ci/run_example_tests.sh:49]: Let's start small, and  use what we need for CI first. Non-used test code / script usually get out of date and stop working.

File: .azure-pipelines/pipelines.yml
Comment (lizan) [reply .azure-pipelines/pipelines.yml:45]: Also I'm not sure if we need docker-in-docker, just use docker seems fine?

Comment (phlax) [reply .azure-pipelines/pipelines.yml:45]: for ci - mostly yep it doesnt matter - although it is quite handy for the env to be clean to test the setup - ie user/perms/umask etc

running locally it matters if you are running docker already and dont want lots of new images and/or the script to prune anything you currently have in your docker env

Comment (phlax) [reply .azure-pipelines/pipelines.yml:45]: hmm

actually it does matter specifically for the umask part of the test

if the repo is already checked out and the umask is not set then these tests would pass without really testing the permissions issue

we could get around that by settting the umask before checking out the repo

File: ci/run_example_tests.sh
Comment (phlax) [reply ci/run_example_tests.sh:49]: see above comment about umask - as that is relevant here  too

Comment (phlax) [reply ci/run_example_tests.sh:49]: my original plan here was to do:

```
export ENVOY_REPO="-b $(git branch --show-current) $(git remote get-url origin)"
```
i abandoned that for 2 reasons - firstly because the repo is already checked out in `ci`, but also because the git `remote` is (i think at least on azure) an ssh remote  - which wont work without a bit more effort

the problem i now realise is that the requirement to checkout the repo again (and use `d-in-d`) is to ensure that the checked out repo has exactly the permmissions that a user with a strict `umask` would have

Comment (lizan) [reply ci/run_example_tests.sh:49]: Let's not worry about umask for now? Aren't we already copying configs inside docker?

Comment (phlax) [reply ci/run_example_tests.sh:49]: > Let's not worry about umask for now? 

really?

i think it makes sense to test for this properly if the expectation is for this to work

> Aren't we already copying configs inside docker?

yep, but i think the point about adding (these) tests is to ensure that the docs continue to work

you can have whatever bits of the code you like, but it seems to me like we are adding a lot of code to then not test the the very thing the tests were used to fix.

i also think that the ci scripts should run locally - i would not want to run this locally outside of a container.

**Comment by phlax** [comment]
@lizan would you mind triggering azure. I need to see if the tests are passing

File: ci/run_example_tests.sh
Comment (lizan) [reply ci/run_example_tests.sh:49]: @phlax My point here is to keep this PR minimum, let CI test what examples we have today, without adding many stuff, these kcan be in follow up PRs.

re: DinD, I don't personally run scripts in a container and neither most of our CI do. We can add support for that but should be optional.

Comment (phlax) [reply ci/run_example_tests.sh:49]: > @phlax My point here is to keep this PR minimum, let CI test what examples we have today, without adding many stuff, these kcan be in follow up PRs.

ill remove the clone code etc - and make the script work entirely standalone

at risk of sounding repetitive, and having just used these tests extensively to fix this particular problem - im surprised that we dont want to ensure it stays fixed

> re: DinD, I don't personally run scripts in a container and neither most of our CI do. We can add support for that but should be optional.

some things run in `ci/run_envoy_docker.sh` - and in my testing with building docs this  was necessary for the build script to work. Perhaps these are build tasks rather than ci, but im  not sure if that matters

the point about using `d-in-d` tho is 2fold
- we want as clean an env as possible so we are actually testing something like an end user's setup
- running all of these tests builds/creates (not properly namespaced) ~20-30 images. This takes up quite a bit of space, and is pita to sit and pick back out

Comment (lizan) [reply ci/run_example_tests.sh]: > at risk of sounding repetitive, and having just used these tests extensively to fix this particular problem - im surprised that we dont want to ensure it stays fixed

Thanks, yes I understand why it exists and I'm not saying we don't want, but not in this PR. Having a script other than CI checkout doing it having a risk that we're testing examples in master (or other branch that the binary is built being checked out), because the CI checkout is fairly complicated and subject to race (because it always use `pull/<PR>/merge` which is a generated commit).

Comment (lizan) [reply ci/run_example_tests.sh]: If DinD is necessary I'm OK leaving it, but I do need a way to run without it, so I can debug the examples with CI

**Comment by lizan** [comment]
btw please dont force push

File: ci/run_example_tests.sh
Comment (phlax) [reply ci/run_example_tests.sh]: cool - ive removed it for now and shifted the code for it back to my repo - so i can continue testing.

yep, lets consider wrapping this test in a separate pr

**Comment by phlax** [comment]
> btw please dont force push

sorry - i figured its ok on a `wip` pr - im testing against this branch - but i can shift to another if necessary

File: ci/verify_examples.sh
Comment (phlax) [inline ci/verify_examples.sh:240]: remove these

File: ci/verify_examples.sh
Comment (phlax) [inline ci/verify_examples.sh:218]: use --head not -v

**Comment by phlax** [comment]
@lizan i have updated the `verify_examples` script, it now runs standalone, and works on azure

i have tidied the script, and dewipped the PR - i think it should be ready for review

i have left the pipeline where it is for now so that tests are triggered/immediately.

once the script itself has approval, ill move it.

im wondering if the test for configs in `BUILD` needs to be separated out - i think this test probably wants to run early/always

one concern  - which i think is an issue with the example itself or perhaps envoy is that while the `mysql` tests work (db/table created etc) they are triggering protocol errors in stats

you can see the `mysql` issue on travis here https://travis-ci.org/github/phlax/envoy-examples/builds/715797859#L5389 - it happens on azure/locally also

File: ci/verify_examples.sh
Comment (lizan) [inline ci/verify_examples.sh:95]: Shall we have each of these function into a separate script so they can be run individually? How do I debug if one of the example test failed? And verify_examples can just glob in sample directory to run them?

Comment (phlax) [reply ci/verify_examples.sh:95]: > Shall we have each of these function into a separate script so they can be run individually? 

not sure. its defo an issue we need to address - i was un/commenting as i went through  the tests

im wondering if adding an argument - eg `ci/verify_examples.sh jaeger*` - might be a good idea - irrespective of whether all of the tests are actually in the file.

i did think about putting `verify.sh` in each of the example folders - to be more explicit for example creators - but then i thought it was distracting for example users

> How do I debug if one of the example test failed? 

you get a bash traceback with the line numbers - this actually works pretty well i think

> And verify_examples can just glob in sample directory to run them?

not sure if i understand exactly this point.

i think we do want to glob in the actual examples folder to work out what tests *should* exist to enforce that examples need a test

somewhat tangentially i had some ideas to make this script a little more concise/readable:

- remove the local `name` in each test - there is already a usable var in scope
- add some  helper functions like `responds_with`  `responds_with_header` `responds_without_header` - this would make the curl tests more explicit as to what they were testing i think

ill update with  these 2 changes shortly - unless you think its a bad idea

Comment (lizan) [reply ci/verify_examples.sh:95]: > > Shall we have each of these function into a separate script so they can be run individually?
> 
> not sure. its defo an issue we need to address - i was un/commenting as i went through the tests
> 
> im wondering if adding an argument - eg `ci/verify_examples.sh jaeger*` - might be a good idea - irrespective of whether all of the tests are actually in the file.
> 
> i did think about putting `verify.sh` in each of the example folders - to be more explicit for example creators - but then i thought it was distracting for example users

How it would distracting? You can have verify_examples.sh run all verify.sh?
Then if I want to run only one example, I can go to that and run verify.sh there.

**Comment by phlax** [comment]
@lizan i have split the tests into `verify.sh` scripts in the respective examples folders

i have also added the helper scripts mentioned, and made some small improvements to the error handling - partly to compensate for moving the tests out of the scope of the error handling in `verify_examples.sh`

verify_examples now accepts a glob  arg so you can run it with individual tests, and the tests should be runnable directly from the example folders also

this script probs needswork - but im not sure about whether the other test - the one for build configs - needs to  be separated or at least runnable separately

i ran the mysql test locally and can confirm that the protocol errors occur outside of these tests when run manually

once we are certain of the format of the tests i can write a small dev doc - preferrably in  a subsequent pr

**Comment by phlax** [comment]
...also - is it ok to rebase/force push as i think the conflict is blocking azure

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:116]: use `'`

File: examples/fault-injection/verify.sh
Comment (phlax) [inline examples/fault-injection/verify.sh:35]: add err message

File: examples/verify-common.sh
Comment (lizan) [inline examples/verify-common.sh:115]: perfer `[[` everywhere: https://google.github.io/styleguide/shellguide.html#s6.3-tests

File: examples/mysql/verify.sh
Comment (phlax) [inline examples/mysql/verify.sh:17]: grep test

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:26]: move after `snooze`

File: examples/cors/verify.sh
Comment (phlax) [inline examples/cors/verify.sh:31]: untabify

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:88]: remove `curl_command`

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:60]: add `path`

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:28]: add `path`

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:73]: add `arg`

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:108]: remove `curl_command`

**Comment by lizan** [comment]
Can you resolve the conflicts and make CI test with newly built image?

File: examples/verify-common.sh
Comment (phlax) [inline examples/verify-common.sh:24]: i think this is unneccessary - the problem was that the  images werent tagged - ill update...

**Comment by sunjayBhatia** [comment]
cc @envoyproxy/windows-dev might be interesting to get Windows coverage of these as well for more confidence etc.

**Comment by phlax** [comment]
@lizan are there any further changes required to land this ?

File: .azure-pipelines/pipelines.yml
Comment (lizan) [inline .azure-pipelines/pipelines.yml:187]: Can you split this to a bash file?

**Review by lizan — COMMENTED** [review]
Thanks, this looks great.

File: .azure-pipelines/pipelines.yml
Comment (phlax) [reply .azure-pipelines/pipelines.yml]: moved to do_ci.sh - lmk if you want it somewhere else

**Comment by phlax** [comment]
@lizan should be ready for review again

**Review by lizan — APPROVED** [review]

**Pull request merged**
